### PR TITLE
Mobile cards carousel

### PR DESCRIPTION
**Pull request by DXCanas**
## Summary

Running into some issues with responsiveWindow - doesn't seem to be working on the learn page when the carousel is on it. Debugging, but let me know what you all think so far.

Notes:
- No "View More" Button yet, as described in the mocks.
- Needs #1898

mobilecarousel

### CODE REVIEW

**Comment by DXCanas** [comment]
Debugging Notes:
For whatever reasons, `<learn-page>`'s `windowSize` data attribute stops updating whenever carousel comes is rendered. It doesn't seem like `responsiveWindow`'s `beforeDestroy` is called in the context of the learn page at any point (though it is for the cards any time their wrapper component changes). So the even listener that `responsiveWindow` creates should still be there. Not sure what's going. Going to continue the debugging tomorrow.

**Comment by DXCanas** [comment]
@rtibbles Tweaking the logic to use a new subset of the contents array like we discussed breaks the carousel as far as I can tell. I agree that it needs cleanup and that it's too fragile right now, but do you think this could wait until after #1898, #1911, #1792 and this PR are merged so that I can do a proper refactor?

**Comment by rtibbles** [comment]
Yep!

File: kolibri/core/assets/src/mixins/responsive-window.js [hunk 1]
```diff
@@ -160,7 +160,7 @@ function addWindowListener(cb) {
 }
 
 function removeWindowListener(cb) {
-  windowListeners.pop(cb);
+  windowListeners.splice(windowListeners.indexOf(cb), 1);
```
Comment (rtibbles) [inline kolibri/core/assets/src/mixins/responsive-window.js:163]: :+1:

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 2]
```diff
@@ -24,11 +24,23 @@
   import PDFobject from 'pdfobject';
   import ScreenFull from 'screenfull';
   import iconButton from 'kolibri.coreVue.components.iconButton';
+  // import PDFjs from 'file-loader?name=[path][name].[ext]!pdf.js';
```
Comment (rtibbles) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: This appears to have been erroneously committed?

File: kolibri/plugins/document_pdf_render/package.json [hunk 3]
```diff
@@ -3,6 +3,8 @@
   "description": "Document/PDF Render Plugin for Kolibri",
   "private": true,
   "dependencies": {
-    "pdfobject": "2.0.201604172"
+    "pdf.js": "https://github.com/mozilla/pdf.js",
```
Comment (rtibbles) [inline kolibri/plugins/document_pdf_render/package.json]: Likewise.

File: kolibri/plugins/document_pdf_render/yarn.lock [hunk 4]
```diff
@@ -2,6 +2,14 @@
 # yarn lockfile v1
```
Comment (rtibbles) [inline kolibri/plugins/document_pdf_render/yarn.lock]: As above.

File: kolibri/plugins/learn/assets/src/views/content-card-carousel/index.vue [hunk 5]
```diff
@@ -82,8 +82,8 @@
   const gutterWidth = 20;
 
   export default {
+    name: 'contentCardCarousel',
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/content-card-carousel/index.vue]: :+1:

File: kolibri/plugins/learn/assets/src/views/content-card-grid/index.vue [hunk 6]
```diff
@@ -80,6 +76,7 @@
       genLink: {
         type: Function,
         validator(value) {
+          // callback: function(contentId, contentKind)
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/content-card-grid/index.vue]: Clean up.

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue [hunk 7]
```diff
@@ -34,12 +34,12 @@
 
 <script>
 
-  import * as CoreConstants from 'kolibri.coreVue.vuex.constants';
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/content-card/index.vue]: No other instances, this is a clean clean up!

File: kolibri/plugins/learn/assets/src/views/explore-page/index.vue [hunk 8]
```diff
@@ -13,19 +13,7 @@
...
-          :key="content.id"
-          :title="content.title"
-          :thumbnail="content.thumbnail"
-          :kind="content.kind"
-          :progress="content.progress"
-          :link="genLink(content)"/>
-      </template>
-
-    </content-card-grid>
+    <content-card-grid :contents="contents" :gen-link="genLink" v-if="contents.length" />
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/explore-page/index.vue]: Delegating the card rendering to the content grid definitely seems more appropriate.

**Comment by DXCanas** [comment]
I had initially rebased on the wrong branch :sweat_smile:

**Comment by rtibbles** [comment]
Seems to work OK. There is the outstanding issue that now the cards are very high for the screen (on a mobile screen, you can only see one card on the screen at a time) - not sure what the answer here is @indirectlylit @khangmach, while also dealing with the aspect ratio issue?

image

**Comment by khangmach** [comment]
@rtibbles @DXCanas 

Well, it would be ideal if we could fit two 16:9 cards side by side on mobile, but I believe that was discussed early on with David as not possible (too much leftover margin & text gets too truncated). If we can't do the above, the alternative was, for each card, to preserve the aspect ratio but also go full width, shown here. The card is tall and takes up a good majority of the screen, which I think is okay for the short term if we only show 2 cards + view more button for each section. 

Sure, it looks ugly, but users will still scroll downward for more content. Also, they won't have to scroll downwards too much because of the view more buttons. Something I've noticed with these thumbnails is that they provide less meaningful information than the thumbnails. The thumbnails are quite arbitrary in what they depict, like a random snapshot of something happening in an exercise or video. If anything it just works to affirm the relevancy and context of the content title (math thumbnail/math card title). 

For the future, I am okay with:
1. Veering off the holy aspect ratio of 16:9 if we can fit more content in the viewport (David's original implementation above at the very top of this PR) 
2. if we can find a way to fit two side by side that would be great. This would require a card redesign that preserves 16:9 for thumbnail only & gets rid of the white card background to accommodate smooth transitions from text to thumbnail (if a title of a video is short, there's still a ton of white space due to the card's aspect ratio)

**Comment by indirectlylit** [comment]
Just wanted to reiterate the simple change I recommended [earlier]((https://github.com/learningequality/kolibri/pull/1898#issuecomment-319468934)): keep the text area a fixed height. It keeps the thumbnail aspect ratio fixed at 16:9 and saves a bit of the wasted space.

It's the difference between this:

image

And this:

image

**Comment by rtibbles** [comment]
I think this is a good minimal fix, having the extra white space seems like a very inefficient use of space.

**Comment by DXCanas** [comment]
Not sure my preference yet, but I'd also like to point out that the text scales on a truly mobile screen, if I recall

**Comment by indirectlylit** [comment]
> text scales on a truly mobile screen

I'm not sure what you mean by this. Can you post a screenshot?

**Comment by DXCanas** [comment]
Initially, I thought that mobile (User Agents or Otherwise) was handling some text scaling to make it look less awkward:
localhost-8000-learn- iphone 5

...But it looks like that's what it looks like on slim displays, because if we bump it up to a wider-screened phone (like the Nexus 5X) the issue you mentioned re-occurs:
localhost-8000-learn- nexus 5x

@khangmach Because of the way the cards have changed in their mobile implementation, it's more feasible than I thought. Would the behavior be to show 2 cards from  one breakpoint to another, and at the smallest breakpoint to onlly show 1?

Do you think this is preferable to the changes outlined by @indirectlylit ?

**Comment by DXCanas** [comment]
Going with @indirectlylit's suggestion

**Comment by DXCanas** [comment]
Widest breakpoint that uses mobile carousel, @khangmach @indirectlylit 
localhost-8000-learn-

File: kolibri/plugins/learn/assets/src/views/learn-page/index.vue [hunk 9]
```diff
@@ -51,9 +64,46 @@
...
+        if (this.isMobile) {
+          return contentCardGrid;
+        }
+        return contentCardCarousel;
+      },
+      trimmedResume() {
+        if (this.isMobile) {
+          return this.recommendations.resume.slice(0, mobileCardNumber);
+        }
+        return this.recommendations.resume.slice(0, mobileCardNumber);
```
Comment (christianmemije) [inline kolibri/plugins/learn/assets/src/views/learn-page/index.vue]: This is returning the same thing that is returned within the if statement.

**Comment by indirectlylit** [comment]
me gusta

File: kolibri/plugins/learn/assets/src/views/learn-page/index.vue (hunk 9 above)
Comment (rtibbles) [reply kolibri/plugins/learn/assets/src/views/learn-page/index.vue]: @DXCanas let's clean this up before merge.

**Pull request merged**